File: sdk_v2/js/src/session.ts
import { FlErrorCode, isFoundryLocalError } from "./detail/errors.js";
import {
  type NativeAudioSession,
  type NativeChatSession,
  type NativeEmbeddingsSession,
  type NativeSession,
  getAddon,
} from "./detail/native.js";
import type { Item } from "./items.js";
// Public `Session` / `ChatSession` classes.
//
// Surface:
//   * `Session` is an abstract base — not directly constructible. Users
//     construct a modality-specific subclass (`ChatSession` today;
//     `AudioSession` / `EmbeddingsSession` later). The base carries the
//     shared inference plumbing (processRequest / processStreamingRequest
//     / setOptions / dispose).
//   * `new ChatSession(model)` — construct directly from a loaded `Model`.
//     The native ctor calls `flSession_Create` synchronously.
//   * `session.processRequest(request) -> Promise<Response>` (non-streaming)
//   * `session.processStreamingRequest(request, { signal? }) ->
//     AsyncIterable<Item>`
//     `AbortSignal` cancellation maps the native `OperationCancelled`
//     rejection to a Web/Node-standard `AbortError`.
//   * `ChatSession` adds turn tracking and tool definitions.
//   * `setOptions(...)`, `dispose()`, `Symbol.dispose`.
import type { IModel } from "./imodel.js";
import { Model, unwrapNativeModel } from "./model.js";
import { type Request, type RequestOptions, unwrapNativeRequest } from "./request.js";
import type { Response } from "./response.js";

/** Options accepted by streaming Session APIs. */
export interface StreamOptions {
  /**
   * Optional cancellation signal. When the signal aborts, the underlying
   * `Request` is cancelled and the async iterator rejects with an `Error`
   * whose `name === "AbortError"` (mirroring the Web/Node standard).
   */
  readonly signal?: AbortSignal;
}

/**
 * Result of {@link Session.processStreamingRequest}. Iterable for streaming
 * items, with a `response` promise that resolves to the terminal `Response`
 * once the native call completes — carrying stop reason, usage, and any
 * non-streamed items (e.g. the final aggregated text item).
 *
 * `response` settles after the iterator finishes draining. It rejects with
 * the same error the iterator would throw (including `AbortError` when the
 * stream is cancelled, and `OperationCancelled` when the consumer breaks
 * early without an `AbortSignal`).
 */
export interface StreamingResponse extends AsyncIterable<Item> {
  readonly response: Promise<Response>;
}

function modelToNativeChatSession(model: IModel): NativeChatSession {
  if (!(model instanceof Model)) {
    throw new TypeError("ChatSession: expected a Model as the first argument");
  }
  const nativeModel = unwrapNativeModel(model);
  return new (getAddon().ChatSession)(nativeModel);
}

function modelToNativeEmbeddingsSession(model: IModel): NativeEmbeddingsSession {
  if (!(model instanceof Model)) {
    throw new TypeError("EmbeddingsSession: expected a Model as the first argument");
  }
  // Validate task in JS BEFORE constructing the native session. Matches the
  // C# (`ValidateTask`) and Python (`__init__` precheck) conventions: a
  // wrong-task model surfaces a `TypeError` rather than a native
  // `FoundryLocalError`, and the check works whether or not the model has
  // been loaded yet (the native ctor would fail later either way).
  const task = model.info.task;
  if (task !== "embeddings") {
    throw new TypeError(`EmbeddingsSession requires a model with task 'embeddings', but got '${task ?? "(unset)"}'.`);
  }
  const nativeModel = unwrapNativeModel(model);
  return new (getAddon().EmbeddingsSession)(nativeModel);
}

function modelToNativeAudioSession(model: IModel): NativeAudioSession {
  if (!(model instanceof Model)) {
    throw new TypeError("AudioSession: expected a Model as the first argument");
  }
  // JS-side task validation, same rationale as `modelToNativeEmbeddingsSession`.
  const task = model.info.task;
  if (task !== "automatic-speech-recognition") {
    throw new TypeError(
      `AudioSession requires a model with task 'automatic-speech-recognition', but got '${task ?? "(unset)"}'.`,
    );
  }
  const nativeModel = unwrapNativeModel(model);
  return new (getAddon().AudioSession)(nativeModel);
}

/**
 * Drive a native streaming session and yield each item to the consumer.
 * Handles backpressure (the JS-side queue grows; the native TSFN backpressure
 * caps producer-side queueing), abort signal wiring, error mapping, and
 * deterministic cleanup on early break.
 *
 * The native call starts eagerly so the returned `response` promise is
 * meaningful even if the caller never iterates (e.g. awaits `.response`
 * directly). The promise settles only after the consumer has fully drained
 * the iterator, mirroring native finalize-on-drain semantics.
 */
function streamItems(
  native: NativeSession,
  request: Request,
  signal: AbortSignal | undefined,
): StreamingResponse {
  const queue: Item[] = [];
  let waiter: (() => void) | null = null;
  let done = false;
  let nativeError: unknown = null;

  const wake = (): void => {
    if (waiter !== null) {
      const w = waiter;
      waiter = null;
      w();
    }
  };

  const onAbort = (): void => {
    try {
      request.cancel();
    } catch {
      // Cancel is best-effort; the request may already be complete.
    }
  };
  if (signal !== undefined) {
    signal.addEventListener("abort", onAbort, { once: true });
  }

  const mapError = (err: unknown): unknown => {
    if (
      signal?.aborted === true &&
      isFoundryLocalError(err) &&
      err.code === FlErrorCode.OperationCancelled
    ) {
      (err as { name: string }).name = "AbortError";
    }
    return err;
  };

  let responseResolve!: (resp: Response) => void;
  let responseReject!: (err: unknown) => void;
  const responsePromise = new Promise<Response>((resolve, reject) => {
    responseResolve = resolve;
    responseReject = reject;
  });
  // Pre-attach a no-op handler so callers who never read `.response` don't
  // see an unhandled rejection. The original promise still rejects when
  // awaited.
  responsePromise.catch(() => {});

  if (signal?.aborted === true) {
    const err = makeAbortError("Stream aborted before start");
    nativeError = err;
    done = true;
    responseReject(err);
  } else {
    const nativeReq = unwrapNativeRequest(request);
    const onItem = (item: unknown): void => {
      // Native may deliver items synchronously during startup before the
      // iterator reaches its first await. That's safe: iterate() always
      // drains queued items before installing a waiter, so no wake is needed
      // when the consumer is not yet parked.
      queue.push(item as Item);
      wake();
    };
    native.processStreamingRequest(nativeReq, onItem).then(
      (resp: unknown) => {
        done = true;
        responseResolve(resp as Response);
        wake();
      },
      (err: unknown) => {
        const mapped = mapError(err);
        nativeError = mapped;
        done = true;
        responseReject(mapped);
        wake();
      },
    );
  }

  async function* iterate(): AsyncGenerator<Item> {
    try {
      while (true) {
        const next = queue.shift();
        if (next !== undefined) {
          yield next;
          continue;
        }
        if (done) {
          if (nativeError !== null) {
            throw nativeError;
          }
          return;
        }
        await new Promise<void>((resolve) => {
          waiter = resolve;
        });
      }
    } finally {
      if (signal !== undefined) {
        signal.removeEventListener("abort", onAbort);
      }
      if (!done) {
        // Consumer broke out early — cancel the request and let the native
        // call settle so the response promise observers don't hang. The
        // response promise will reject with the cancellation error via the
        // native `.then` error handler.
        try {
          request.cancel();
        } catch {
          // ignore
        }
        while (!done) {
          await new Promise<void>((resolve) => {
            waiter = resolve;
          });
        }
      }
    }
  }

  return {
    [Symbol.asyncIterator]: () => iterate(),
    response: responsePromise,
  };
}

function makeAbortError(message: string): Error {
  const err = new Error(message);
  err.name = "AbortError";
  return err;
}

export abstract class Session {
  // `protected` (not `#private`) so subclasses can downcast for
  // modality-specific native methods without a second field/storage slot.
  protected readonly native: NativeSession;

  protected constructor(native: NativeSession) {
    this.native = native;
  }

  /**
   * Run inference for `request`. Resolves with a `Response` snapshot when
   * generation completes. The full output is materialised before resolution
   * — use {@link processStreamingRequest} to consume items incrementally.
   *
   * Rejects with a `FoundryLocalError` on native failure. Calling
   * `request.cancel()` from another async context causes this promise to
   * reject with `code === FlErrorCode.OperationCancelled`.
   */
  async processRequest(request: Request): Promise<Response> {
    const nativeReq = unwrapNativeRequest(request);
    return (await this.native.processRequest(nativeReq)) as Response;
  }

  /**
   * Run inference for `request`, yielding each `Item` produced by the model
   * as it streams. The returned value is an `AsyncIterable<Item>` that can
   * be consumed with `for await (const item of session.processStreamingRequest(req)) { ... }`,
   * and also exposes a `response` promise that resolves to the terminal
   * `Response` (stop reason, usage, aggregate text item, etc.) once the
   * native call completes.
   *
   * Cancellation: pass `{ signal }`; aborting the signal cancels the native
   * request and causes the iterator to throw an `Error` with
   * `name === "AbortError"`. Breaking out of the `for await` loop also
   * cancels the underlying request.
   *
   * Non-cancellation failures throw a `FoundryLocalError`.
   */
  processStreamingRequest(request: Request, options?: StreamOptions): StreamingResponse {
    return streamItems(this.native, request, options?.signal);
  }

  /** Apply session-level options that persist across `processRequest()` calls. */
  setOptions(options: RequestOptions): this {
    this.native.setOptions(options);
    return this;
  }

  /** True once `dispose()` has run. */
  get disposed(): boolean {
    return this.native.isDisposed();
  }

  /**
   * Release the underlying native session handle. Idempotent. After
   * disposal, every other instance method rejects (async) or throws (sync)
   * with `FoundryLocalError` / `code === FlErrorCode.InvalidUsage`.
   */
  dispose(): void {
    this.native.dispose();
  }

  [Symbol.dispose](): void {
    this.dispose();
  }
}

/** A tool definition exposed to {@link ChatSession.addToolDefinition}. */
export interface ToolDefinition {
  readonly name: string;
  readonly description: string;
  readonly jsonSchema: string;
}

export class ChatSession extends Session {
  /**
   * Construct a `ChatSession` bound to `model`. The model must be loaded
   * (call `await model.load()` first) or have been previously loaded. The
   * chat session additionally tracks conversation history and tool
   * definitions across `processRequest()` calls.
   *
   * Throws `TypeError` if `model` is not a `Model` instance, and a
   * `FoundryLocalError` if the native session cannot be created.
   */
  constructor(model: IModel) {
    super(modelToNativeChatSession(model));
  }

  // Typed accessor for chat-only methods on the native handle. The base
  // stores the same handle as a `NativeSession`; the cast is safe because
  // `ChatSession` always constructs a `NativeChatSession`.
  get #nativeChat(): NativeChatSession {
    return this.native as NativeChatSession;
  }

  /**
   * Register a tool definition available to the model for the rest of the
   * session. Mirrors `foundry_local::ChatSession::AddToolDefinition`.
   */
  addToolDefinition(definition: ToolDefinition): this {
    this.#nativeChat.addToolDefinition(definition);
    return this;
  }

  /**
   * Remove a previously registered tool definition by name. Returns `true`
   * if a tool with that name existed and was removed, `false` otherwise.
   * Mirrors `foundry_local::ChatSession::RemoveToolDefinition`.
   */
  removeToolDefinition(name: string): boolean {
    return this.#nativeChat.removeToolDefinition(name);
  }

  /** Number of completed turns. */
  get turnCount(): number {
    return this.#nativeChat.turnCount();
  }

  /**
   * Rewind the last `count` turns: removes their input messages and
   * assistant replies from history and resets the generator.
   */
  undoTurns(count: number): void {
    this.#nativeChat.undoTurns(count);
  }
}

/**
 * Inference session for text-embedding models.
 *
 * Embeddings sessions are stateless and one-shot. Use the inherited
 * {@link Session.processRequest} method with `TextItem` inputs; the
 * response contains one `TensorItem` per input holding the embedding
 * vector.
 *
 * Streaming is not supported for embeddings — calling
 * {@link Session.processStreamingRequest} will fail at the native
 * boundary because the native `EmbeddingsSession` ObjectWrap does not
 * register a `processStreamingRequest` method.
 *
 * Task validation runs in the JS layer before the native ctor: passing a
 * non-embeddings model throws `TypeError` regardless of whether the
 * model has been loaded.
 */
export class EmbeddingsSession extends Session {
  constructor(model: IModel) {
    super(modelToNativeEmbeddingsSession(model));
  }
}

/**
 * Inference session for automatic-speech-recognition models.
 *
 * Accepts `AudioItem` input (from URI or in-memory PCM bytes) and produces
 * `TextItem` output containing the transcription. For live streaming
 * input, push successive `Item.bytes(chunk)` PCM frames onto an
 * `ItemQueue`, add the queue to the request, and call `markFinished()`
 * when the input stream ends.
 *
 * Task validation runs in the JS layer before the native ctor: passing a
 * non-ASR model throws `TypeError` regardless of whether the model has
 * been loaded.
 */
export class AudioSession extends Session {
  constructor(model: IModel) {
    super(modelToNativeAudioSession(model));
  }
}

File: sdk_v2/cs/src/Session.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

using System.Threading.Channels;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.AI.Foundry.Local.Detail.Interop;
using Microsoft.AI.Foundry.Local.Detail.Native;

using NativeSession = Microsoft.AI.Foundry.Local.Detail.Native.Session;

/// <summary>
/// Base session wrapping the native inference session.
/// Provides request processing, streaming, options, and tool definitions.
/// Use <see cref="ChatSession"/> or <see cref="AudioSession"/> for task-specific validation.
/// </summary>
public abstract class Session : IDisposable
{
    private readonly NativeSession _session = null!;
    private readonly ManagerLifetime _managerLifetime = null!;
    private readonly ManagerLifetime.SessionRegistration _managerRegistration = null!;
    private readonly SessionOperationGate _operationGate = new();
    private readonly OwnershipSlot<StreamingOperation> _streamSlot = new();
    private FlStreamingCallback? _nativeStreamingCallback;
    private int _disposed;

    /// <summary>
    /// Create a session from a loaded model. Subclasses should validate the model task before calling this.
    /// </summary>
    protected Session(IModel model)
    {
        var concrete = (Model)model;
        using var managerLease = concrete.AcquireManagerLease(trackReentrancy: true);
        try
        {
            _session = new NativeSession(concrete.NativeModel);
            _managerLifetime = concrete.NativeLifetime;
            _managerRegistration = concrete.NativeLifetime.RegisterSession(this);
        }
        catch
        {
            _session?.Dispose();
            throw;
        }
    }

    /// <summary>
    /// Set session-level inference options. These apply to all subsequent
    /// <see cref="ProcessRequestAsync"/> calls unless overridden per-request.
    /// </summary>
    /// <returns>This session (fluent).</returns>
    public Session SetOptions(RequestOptions options)
    {
        Detail.Throw.IfNull(options);

        using var operation = _operationGate.Acquire(this);
        using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);

        Api.Root.CreateKeyValuePairs(out var kvpPtr);

        try
        {
            foreach (var kvp in options.ToDictionary())
            {
                Api.Root.AddKeyValuePair(kvpPtr, kvp.Key, kvp.Value);
            }

            _session.SetOptions(kvpPtr);
        }
        finally
        {
            Api.Root.KeyValuePairsRelease(kvpPtr);
        }

        return this;
    }

    /// <summary>
    /// Enable or disable streaming mode. When enabled, a native streaming callback is installed
    /// on the session. Use <see cref="ProcessStreamingRequestAsync"/> to receive items as they are generated.
    /// The callback remains installed until disabled or the session is disposed.
    /// </summary>
    /// <returns>This session (fluent).</returns>
    public Session SetStreaming(bool enabled)
    {
        using var operation = _operationGate.Acquire(this);
        using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);

        if (enabled && _nativeStreamingCallback == null)
        {
            _nativeStreamingCallback = (FlStreamingCallbackData data, IntPtr userData) =>
            {
                var stream = _streamSlot.Value;

                if (stream == null)
                {
                    return 0;
                }

                bool errored = false;

                try
                {
                    if (data.ItemQueue != IntPtr.Zero)
                    {
                        while (Api.Item.QueueTryPop(data.ItemQueue, out var itemPtr))
                        {
                            // Ownership transfers to the channel consumer who disposes it
#pragma warning disable IDISP001
                            var item = Item.FromNative(itemPtr, ownsHandle: true);
#pragma warning restore IDISP001
                            if (!stream.Channel.Writer.TryWrite(item))
                            {
                                item.Dispose();
                            }
                        }
                    }
                }
                catch (Exception ex)
                {
                    errored = true;
                    stream.Channel.Writer.TryComplete(
                        new FoundryLocalException("Error processing streaming callback data.", ex));
                }

                return errored || stream.Cts.IsCancellationRequested ? 1 : 0;
            };

            _session.SetStreamingCallback(_nativeStreamingCallback);
        }
        else if (!enabled && _nativeStreamingCallback != null)
        {
            _session.SetStreamingCallback(null);
            _nativeStreamingCallback = null;
        }

        return this;
    }

    /// <summary>
    /// Process a request and return the complete response.
    /// </summary>
    public async Task<Response> ProcessRequestAsync(Request request, CancellationToken ct = default)
    {
        Detail.Throw.IfNull(request);
        var operation = _operationGate.Acquire(this, assignCurrentThread: false);

        return await Task.Run(() =>
        {
            operation.SetCurrentThread();
            try
            {
                ct.ThrowIfCancellationRequested();
                using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);
                var responsePtr = _session.ProcessRequest(request.Ptr);
                return new Response(responsePtr);
            }
            finally
            {
                operation.ClearCurrentThread();
                operation.Dispose();
            }
        }, CancellationToken.None).ConfigureAwait(false);
    }

    /// <summary>
    /// Process a request with streaming. Returns a <see cref="StreamingResponse"/> whose async
    /// iterator yields <see cref="Item"/>s as they are produced and whose
    /// <see cref="StreamingResponse.FinalResponse"/> resolves to the terminal
    /// <see cref="Response"/> (carrying <see cref="FinishReason"/>, usage, and any aggregated
    /// items) after the iterator drains.
    ///
    /// Requires <see cref="SetStreaming"/> to have been called with <c>true</c>.
    /// Concurrent streaming requests on the same session are not supported.
    ///
    /// The caller MUST either await <see cref="StreamingResponse.FinalResponse"/> (and dispose the
    /// returned <see cref="Response"/>) or <c>await using</c> the <see cref="StreamingResponse"/>
    /// to avoid leaking the native response handle.
    /// </summary>
    /// <exception cref="InvalidOperationException">
    /// Thrown if streaming has not been enabled via <see cref="SetStreaming"/>, or if another
    /// streaming request is already in flight on this session (concurrent streaming requests
    /// on the same session are not supported).
    /// </exception>
    public StreamingResponse ProcessStreamingRequestAsync(Request request, CancellationToken ct = default)
    {
        Detail.Throw.IfNull(request);

#pragma warning disable IDISP001 // Ownership transfers to StreamingOperation and then StreamingResponse.
#pragma warning disable IDISP016 // The operation is disposed only on the throwing precondition branch.
    var operation = _operationGate.TryAcquire(this, assignCurrentThread: false);
    if (operation == null)
    {
        throw new InvalidOperationException("Concurrent streaming requests on the same session are not supported.");
    }

        if (_nativeStreamingCallback == null)
        {
            operation.Dispose();
            throw new InvalidOperationException(
                "Streaming not enabled. Call SetStreaming(true) before ProcessStreamingRequestAsync.");
        }

        var channel = Channel.CreateUnbounded<Item>(
            new UnboundedChannelOptions
            {
                SingleWriter = true,
                SingleReader = true,
                AllowSynchronousContinuations = true,
            });

        var cts = CancellationTokenSource.CreateLinkedTokenSource(ct);
        var stream = new StreamingOperation(channel, cts, operation);
        if (!_streamSlot.TrySet(stream))
        {
            stream.Dispose();
            throw new InvalidOperationException("A streaming response still owns this session.");
        }

        var tcs = new TaskCompletionSource<Response>(TaskCreationOptions.RunContinuationsAsynchronously);

        _ = Task.Run(() =>
        {
            operation.SetCurrentThread();
            try
            {
                try
                {
                    using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);
                    var responsePtr = _session.ProcessRequest(request.Ptr);

                    // Capture cancellation before channel completion, which can synchronously run cleanup.
                    var wasCancelledBeforeReturn = cts.IsCancellationRequested;

                    channel.Writer.TryComplete();

                    if (wasCancelledBeforeReturn)
                    {
                        Api.Inference.ResponseRelease(responsePtr);
                        tcs.TrySetCanceled(cts.Token);
                    }
                    else
                    {
#pragma warning disable IDISP004 // Ownership transferred to FinalResponse consumer (or DisposeAsync).
                        tcs.TrySetResult(new Response(responsePtr));
#pragma warning restore IDISP004
                    }
                }
                catch (OperationCanceledException)
                {
                    channel.Writer.TryComplete();
                    tcs.TrySetCanceled(cts.Token);
                    return;
                }
                catch (Exception ex)
                {
                    var wrapped = new FoundryLocalException("Error executing streaming request.", ex);
                    channel.Writer.TryComplete(wrapped);
                    tcs.TrySetException(wrapped);
                    return;
                }
            }
            finally
            {
                operation.ClearCurrentThread();
                stream.MarkProducerCompleted();
                if (_operationGate.IsClosed)
                {
                    ReleaseStreamingOperation(stream);
                }
            }
        }, CancellationToken.None);

        if (_operationGate.IsClosed)
        {
            cts.Cancel();
        }

        return new StreamingResponse(this, stream, tcs);
    #pragma warning restore IDISP016
    #pragma warning restore IDISP001
    }

    internal void ReleaseStreamingOperation(StreamingOperation stream)
    {
        if (_streamSlot.ClearIfOwned(stream))
        {
            stream.Release();
        }
    }

    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    ~Session()
    {
        try
        {
            Dispose(false);
        }
        catch
        {
        }
    }

    protected virtual void Dispose(bool disposing)
    {
#pragma warning disable IDISP023 // Finalization must close operations and release session before manager registration.
        if (!_operationGate.BeginClose())
        {
            return;
        }

        Interlocked.Exchange(ref _disposed, 1);

        try
        {
            var stream = _streamSlot.Value;

            stream?.Cancel();
            stream?.WaitForProducer();
            if (stream != null)
            {
                ReleaseStreamingOperation(stream);
            }
        }
        finally
        {
            try
            {
                _operationGate.WaitForOperations();
            }
            finally
            {
                try
                {
                    _session?.Dispose();
                }
                finally
                {
                    _managerRegistration?.Dispose();
                }
            }
        }
#pragma warning restore IDISP023
    }

    protected T ExecuteNative<T>(Func<NativeSession, T> operation)
    {
        using var admission = _operationGate.Acquire(this);
        using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);
        return operation(_session);
    }

    protected void ExecuteNative(Action<NativeSession> operation)
    {
        using var admission = _operationGate.Acquire(this);
        using var managerLease = _managerLifetime.Acquire(this, trackReentrancy: true);
        operation(_session);
    }

    protected void ThrowIfDisposed()
    {
        Detail.Throw.IfDisposed(Volatile.Read(ref _disposed) != 0, this);
    }

    internal sealed class StreamingOperation : IDisposable
    {
        private readonly SessionOperationGate.Operation _operation;
        private readonly TaskCompletionSource<bool> _producerCompleted =
            new(TaskCreationOptions.RunContinuationsAsynchronously);
        private int _released;

        internal StreamingOperation(Channel<Item> channel, CancellationTokenSource cts,
                                    SessionOperationGate.Operation operation)
        {
            Channel = channel;
            Cts = cts;
            _operation = operation;
        }

        internal Channel<Item> Channel { get; }
        internal CancellationTokenSource Cts { get; }
        internal Task ProducerTask => _producerCompleted.Task;
        internal void MarkProducerCompleted() => _producerCompleted.TrySetResult(true);
        internal void WaitForProducer() => _producerCompleted.Task.GetAwaiter().GetResult();
        internal void Cancel()
        {
            try { Cts.Cancel(); } catch { }
        }

        internal void Release()
        {
            if (Interlocked.Exchange(ref _released, 1) == 0)
            {
#pragma warning disable IDISP007 // Ownership transferred into StreamingOperation by its constructor.
                _operation.Dispose();
#pragma warning restore IDISP007
            }
        }

        public void Dispose() => Release();
    }
}
